clicache/src/Cache.cpp

```diff
@@ -57,6 +56,21 @@ namespace Apache
         m_typeRegistry = gcnew Apache::Geode::Client::TypeRegistry(this);
       }
 
+      Cache::~Cache() { //Destructor - deterministic
+        if (!_disposed) {
+          //Clean-up managed resources
+          this->!Cache();
+          _disposed = true;
+          //GC.SuppressFinalize(this) is automatically added here
+          //Base destructor is automatically called too if needed
+        }
+      }
+
+      Cache::!Cache() { //Finalizer - non-deterministic when called by GC
+        //Clean-up unmanaged resources
+        delete m_nativeptr;
```

Comment:
Technically this value is a managed resource. If you are following the example from https://manski.net/2012/01/idisposable-finalizer-and-suppressfinalize/ then it should be deleted in the destructor.

Reply:
The unmanaged resources are deleted in both the destructor (which calls the finalizer) and the finalizer,

Reply:
From my reading of this guide the C++/CLI should delete managed resources in the destructor and free native resources the finalizer.
> Beside unmanaged data, a managed class can also contain managed data, i.e. instances of managed classes implementing IDisposable. Managed data is different from unmanaged data in that it should be disposed in Dispose() but not in the finalizer. This is because instances of managed classes may already have been garbage collected when the finalizer runs.
```C++
ref class DataContainer {
public:
   ...

  ~DataContainer() {
    if (m_isDisposed)
       return;

    delete m_managedData; // dispose managed data
    this->!DataContainer(); // call finalizer
    m_isDisposed = true;
  }

  // Finalizer
  !DataContainer() {
    DataProvider::DeleteUnmanagedData(m_unmanagedData); // free unmanaged data
  }

private:
  bool m_isDisposed;
  IntPtr m_unmanagedData;
  IDisposable^ m_managedData;
};
```

This class technically only has managed data which should be deleted in the dispose/destructor.

The unmanaged data is held in the managed `native_shared_ptr` class.

Reply:
In fact, this class has no need for a finalizer.

Reply:
Good catch. Forgot that m_nativeptr is actually a gcnew object.